Micro-optimize jsStringWithCache()

cdumez opened the pull request:
#### 668e78fa29ae0a403cfc8a34b7973e6965658b00
<pre>
Micro-optimize jsStringWithCache()
<a href="https://bugs.webkit.org/show_bug.cgi?id=255106">https://bugs.webkit.org/show_bug.cgi?id=255106</a>

Reviewed by Geoffrey Garen.

Micro-optimize jsStringWithCache():
- Call JSString::getValueImpl() instead of tryGetValueImpl() on the cached
  JSString. We know we never cache JSRopeStrings.
- Call JSString::create() directly in the slow path instead of jsString()
  to avoid special handling of strings with length 0 and 1. It is unecessary
  since the fast path in jsStringWithCache() already did this special handling.

* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC::jsStringWithCacheSlowCase):
* Source/JavaScriptCore/runtime/JSString.h:
(JSC::jsStringWithCache):

Canonical link: <a href="https://commits.webkit.org/262712@main">https://commits.webkit.org/262712@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd4a6323d5e75ff9a8d57b0e0791e639d83977a8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2277 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2300 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2385 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3204 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2305 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2389 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2359 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2061 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2298 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/2025 "44 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2052 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3062 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2027 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1924 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/1908 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2108 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2056 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3234 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/2184 "Built successfully and passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2073 "Passed tests") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1882 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2344 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2089 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2037 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/529 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/586 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2016 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2395 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2211 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/643 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/JSString.h:
```diff
@@ -1036,22 +1037,23 @@ inline JSString* jsOwnedString(VM& vm, const String& s)
-            return vm.smallStrings.singleCharacterString(static_cast<unsigned char>(singleCharacter));
+    auto& stringImpl = *s.impl();
+    if (length == 1) {
+        auto c = stringImpl[0];
+        if (c <= maxSingleCharacterString)
+            return vm.smallStrings.singleCharacterString(c);
```
You didn’t like the name `singleCharacter`? I probably would have merged the if statement like this:
```
if (auto c = stringImpl[0]; c <= maxSingleCharacterString)
```

cdumez replied:
jsStringWithCache() was inconsistent with the other functions in this file so I tried to fix this while I was touching it.
I'll make the change you're suggesting in this function and others.

cdumez replied:
Fixing via https://github.com/WebKit/WebKit/pull/12800.

geoffreygaren reviewed (approved):
r=me

cdumez committed:
Micro-optimize jsStringWithCache()
https://bugs.webkit.org/show_bug.cgi?id=255106

Reviewed by Geoffrey Garen.

Micro-optimize jsStringWithCache():
- Call JSString::getValueImpl() instead of tryGetValueImpl() on the cached
  JSString. We know we never cache JSRopeStrings.
- Call JSString::create() directly in the slow path instead of jsString()
  to avoid special handling of strings with length 0 and 1. It is unecessary
  since the fast path in jsStringWithCache() already did this special handling.

* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC::jsStringWithCacheSlowCase):
* Source/JavaScriptCore/runtime/JSString.h:
(JSC::jsStringWithCache):

Canonical link: https://commits.webkit.org/262712@main

darinadler reviewed (commented):
Oops I forgot to push the review button and had one remark typed in from back when I reviewed.
